### PR TITLE
update to `prepare_model_for_kbit_training`

### PR DESCRIPTION
**Pull request by mnoukhov**
since `peft` has deprecated `prepare_model_for_int8_training`

also add `use_gradient_checkpointing=args.gradient_checkpointing` to automatically follow the gradient checkpointing choice in training args

For RewardTrainer, this is the workaround to #480 proposed by #694. 

Concurrently @lewtun is working on #726 which adds the `use_gradient_checkpointing` for RewardTrainer. I'm happy to wait until it is merged to merge this.

### CODE REVIEW

**Comment by mnoukhov** [comment]
I've realized this fix actually causes an issue. Calling `model.gradient_checkpointing_enable()` twice leads to the same error as #480. And it is called twice:
1. in `peft`'s `prepare_model_for_kbit_training`  [here](https://github.com/huggingface/peft/blob/0c9354bda98eb7f5348699e23ab752e8dca1e60e/src/peft/utils/other.py#L115C15-L115C44)
2. in huggingface `Trainer'`s inner training loop [here](https://github.com/huggingface/peft/blob/0c9354bda98eb7f5348699e23ab752e8dca1e60e/src/peft/utils/other.py#L115C15-L115C44)

I've made a workaround here in `sft_trainer` but it is ugly and maybe there's a better way. 

To demonstrate the issue, I fixed `stack_llama_2/sft_llama2.py` by making the device map use `Accelerator` so it actually runs on multi-gpu on a single device. Run `accelerator launch --multi_gpu sft_llama2.py` without fix in `sft_trainer` and it will error

**Review by younesbelkada — COMMENTED** [review]
This looks good to me! Thanks for deepdiving and explaining, I left one question, let me know what do you think

**Comment by lewtun** [comment]
FYI @mnoukhov my PR #726 has now been merged so feel free to wrap this one up - thank you 🚀 !

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -145,7 +146,11 @@ def __init__(
                     )
 
                 if getattr(model, "is_loaded_in_8bit", False) or getattr(model, "is_loaded_in_4bit", False):
-                    model = prepare_model_for_int8_training(model)
+                    model = prepare_model_for_kbit_training(
+                        model, use_gradient_checkpointing=args.gradient_checkpointing
+                    )
+
+                    args = dataclasses.replace(args, gradient_checkpointing=False)
```
[the remark this answers is not in the document]
Comment (mnoukhov) [reply trl/trainer/sft_trainer.py:153]: we do want to call `gradient_checkpointing_enable` once, we just don't want to call it twice. We will call it in 'prepare_for_kbit_training` but this change makes sure we don't call it in `Trainer`

[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:153]: Perfect makes sense!

**Comment by younesbelkada** [comment]
@mnoukhov thanks again for your work on this ! Would you be happy to fix the merge conflicts? After that we should be good to merge!

**Comment by mnoukhov** [comment]
Pulled and should be ready to merge!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for this great effort!
